### PR TITLE
GridPlot: grid isosurface was always calculated based on real values

### PR DESCRIPTION
**Pull request by pfebrer**
@tfrederiksen discovered the bug that `imag`, `mod`, `deg_phase`, `rad_phase` representations in 3D were all wrong.

This was basically because I was not modifying the values of the grid before running the `isosurface` method. Therefore, it was calculating **always** isosurfaces based on the real values.

The bug popped out luckily because when trying to calculate isosurfaces for the phase representations, the value was not inside of the range of real values of the grid, so `marching_cubes` complain.

So, message: **3D REPRESENTATIONS OTHER THAN `real` ARE WRONG BEFORE THIS BUG FIX**

This bug also made me realize that `marching_cubes` always converts the grid to `float` before calculating isosurfaces. Therefore it might be helpful to allow an argument in `Grid.isosurface` to let the user specify that they want the isosurface of the imaginary values, the modulus or the phase. What do you think?

PD: Sorry for the bug :(

### CODE REVIEW

**Comment by pfebrer** [comment]
I will add tests for this, don't merge it yet.

File: sisl/viz/plotly/plots/grid.py [hunk 1]
```diff
@@ -512,6 +512,8 @@ def _set_data(self, axes, nsc, interp, trace_name, transforms, represent, cut_va
         elif represent.endswith("phase"):
             values = np.angle(grid.grid, deg=represent.startswith("deg"))
```
Comment (zerothi) [inline sisl/viz/plotly/plots/grid.py:513]: What happens if `represent` isn't any of these?

File: sisl/viz/plotly/plots/grid.py [hunk 2]
```diff
@@ -512,6 +512,8 @@ def _set_data(self, axes, nsc, interp, trace_name, transforms, represent, cut_va
...
 
+        grid.grid = values
+
         #Remove the leftover dimensions
         values = np.squeeze(values)
```
Comment (zerothi) [inline sisl/viz/plotly/plots/grid.py:519]: what about using `plot_func = getattr(self, f"_plot{values.ndim}D")` ?
That would also raise an issue for erroneous data?

**Review by zerothi — CHANGES_REQUESTED** [review]
It looks like `values` isn't defined in all cases?

Secondly, one may be interested in plotting both real+imaginary parts in one plot.
In particular one could then use 4 colors: real positive/negative and imag positive/negative

That is quite useful in some cases.

PS. No worries about bugs, we all make them, and better to fix them. :)

File: sisl/viz/plotly/plots/grid.py (hunk 1 above)
Comment (pfebrer) [reply sisl/viz/plotly/plots/grid.py:513]: Good point, the obvious thing to do here would be
```
else:
    raise ValueError()
```
but since parameters are defined with input fields and input fields parse values before passing them to the plot I think it would be a good idea to raise the error when parsing, to avoid unnecessary calculations (the code would run until here, which might take a while, even though it's clearly deemed to fail). 

Also, raising the error on parse would make life much more easy for GUIs to know if there's a problem with a specific setting, for example, as they could check in real time if the setting is valid or not before trying to update the plot.

By the way, you have made me realise that there should probably be a new error type, something like `SislVizSettingValueError` for the same reason.

File: sisl/viz/plotly/plots/grid.py (hunk 2 above)
Comment (pfebrer) [reply sisl/viz/plotly/plots/grid.py:519]: Done! (not pushed yet)

**Comment by pfebrer** [comment]
> Secondly, one may be interested in plotting both real+imaginary parts in one plot.
> In particular one could then use 4 colors: real positive/negative and imag positive/negative

This can already be done with the general functionality of `MultiplePlot` (https://zerothi.github.io/sisl/docs/latest/visualization/plotly/combining-plots/Intro%20to%20multiple%20plots.html). It allows you to do:

```python
grid.plot(represent=["imag", "real"], varying="represent")
```

This would draw both the imaginary and real contributions in the same plot, while

```python
grid.plot(represent=["imag", "real"], subplots="represent")
```

would create a subplot for each of them (warning, 3D subplots are not possible yet), and

```python
grid.plot(represent=["imag", "real"], animate="represent")
```

would create animation where the first step displays the real part and the second step displays the imaginary part. 

So I'd say it is unnecessary to implement it inside `GridPlot`. It would also make the code more complex. But I'm open to discuss it :)

PD: I don't like the keyword `varying` at all, if you have a better suggestion it would be great!

**Comment by zerothi** [comment]
> > Secondly, one may be interested in plotting both real+imaginary parts in one plot.
> > In particular one could then use 4 colors: real positive/negative and imag positive/negative
> 
> This can already be done with the general functionality of `MultiplePlot` (https://zerothi.github.io/sisl/docs/latest/visualization/plotly/combining-plots/Intro%20to%20multiple%20plots.html). It allows you to do:
> 
> ```python
> grid.plot(represent=["imag", "real"], varying="represent")
> ```
> 
> This would draw both the imaginary and real contributions in the same plot, while
> 
> ```python
> grid.plot(represent=["imag", "real"], subplots="represent")
> ```
> 
> would create a subplot for each of them (warning, 3D subplots are not possible yet), and
> 
> ```python
> grid.plot(represent=["imag", "real"], animate="represent")
> ```
> 
> would create an animation where the first step displays the real part and the second step displays the imaginary part.
> 
> So I'd say it is unnecessary to implement it inside `GridPlot`. It would also make the code more complex. But I'm open to discuss it :)
> 
> PD: I don't like the keyword `varying` at all, if you have a better suggestion it would be great!

Ok! :)

What about `variables`?

**Comment by zerothi** [comment]
Don't know if there are multiple variables, or always 1, in the latter case `variable`?

**Comment by pfebrer** [comment]
```python
grid.plot(setting1=[0,1], setting2=[2,3] varying=["setting1", "setting2"])
```

is allowed. It creates a plot with `(setting1, setting2) = (0, 2)` and another one where `(setting1, setting2) = (1, 3)`.

Also
```python
grid.plot(varying={"setting1": [0,1], "setting2": [2,3]})
```

is equivalent.

So, `variables`? I was looking for some word that makes it clear that they are going to be drawed in the same plot. I find `variables` better than `varying`, but still not ideal :) Also it is important that the name does not clash with names that people could give to settings, and in that respect `variables` seems more dangerous to me. Taking into account that I think I like `variable` better (as an adjective, not as a noun :smile: ).

**Comment by tfrederiksen** [comment]
Have you considered having a single parameter here to control what kind of representation to be produced, e.g. `style : {"single", "multiple", "subplots", "animation"}`? To me this seems more intuitive from the user perspective, but perhaps it presents some limitations?

**Comment by pfebrer** [comment]
Yes, but then it is extremely difficult to determine which are the settings that need to be iterated. Unless you have some idea about it that I didn't think about. You can't check for lists, because there might be settings that accept lists as a value.

**Comment by zerothi** [comment]
Perhaps it would be useful to use tuples or callable in `setting` which are then elements in the varying space.
This is a bit unintuitive but follows the pandas design where list and tuples have different meanings. It would save an argument.

**Comment by pfebrer** [comment]
**Now related to the bug fix:**

While writing the tests I found that maybe the order in which things are applied during the processing of the grid is not optimal. Here is the order:
- Transform the grid (i.e. apply functions such as `np.cos`).
- Cut the grid according to `x_range`, `y_range` and `z_range`.
- Interpolate values to a grid of different shape if `interp` is passed.
- Average dimensions that are not going to be plotted.
- Choose representation (`real`, `imag`, ...)

Could you comment on which order do you think it's best? I'm specially concerned about whether the representation should be chosen at the beggining or at the end.

Maybe it makes sense to allow the user to choose the order of processing?

**Comment by zerothi** [comment]
> Have you considered having a single parameter here to control what kind of representation to be produced, e.g. `style : {"single", "multiple", "subplots", "animation"}`? To me this seems more intuitive from the user perspective, but perhaps it presents some limitations?

I think @tfrederiksen has a point, something like:
```
style={`animate': {"var1": [1, 2, 3], 'style': {"multiple": {"var2": [1, 2, 3]}}}, "multiple": {"var3": [2, 3, 4}}
```
the dictionary levels would be double nested (no animation within an animation. Whereas there could be as many of the others as needed?

1-keyword here I think would be best... I know it is a bit complicated.. But...
```
plot(multiple={"var3": [1, 2, 3]}, animate={"var1": [1, 2, 3]})
```
is already quite verbose, so the chances of this being automated in some way is pretty high.

Alternatively you could create objects which determine how to mix them (they should be hashable, mergeable and iterable), and the contained values could even be grabbed from the `__call__` method?
```
class Animate:
    variable: str
    values: list-like
class Multiple
    variable: str
    values: list-like

var3 = Multiple("var3", [1, 2, 3])
var1 = Animate("var1", [1, 2, 3])
...plot(style=var3 & var1) # meaning both at same time
...plot(style=var3 | var1) # meaning the product of the two combinations
```
Ok, perhaps this last idea isn't as great... Just throwing in some ideas.

**Comment by pfebrer** [comment]
I moved the discussion about multiple plots to https://github.com/zerothi/sisl/issues/340 so that we can focus here on the bug that the PR solves and other things related to `GridPlot` :)

So, in the last push I:
- Moved choosing the representation to the beggining of the processing chain
- Added the `reduce_method` setting, which accepts `{"average", "sum"}` to determine how to reduce the undisplayed dimensions. I did this because sometimes I was annoyed by the plot averaging the axis when I wanted to see the total contribution (to analyze charge accumulation). Complaints suggestions about the setting's name?
- Added tests for all different representations in 1D, 2D and 3D.

Nick: You added the `skip_skimage` part at the top of the `test_grid.py` file. For me it skips tests even though `skimage` is available and `skip_skimage` is set to `pytest.mark.skip(False, reason="scikit-image (skimage) not available")`. Does it work properly for you?

**Comment by zerothi** [comment]
> I moved the discussion about multiple plots to #340 so that we can focus here on the bug that the PR solves and other things related to `GridPlot` :)

Good! Thanks!

> So, in the last push I:
> 
> * Moved choosing the representation to the beggining of the processing chain
> * Added the `reduce_method` setting, which accepts `{"average", "sum"}` to determine how to reduce the undisplayed dimensions. I did this because sometimes I was annoyed by the plot averaging the axis when I wanted to see the total contribution (to analyze charge accumulation). Complaints/suggestions about the setting's name?

Couldn't we re-use the `apply` naming scheme to allow function arguments or str`s to be fetched from `numpy`? That would also allow more functionality without complicating the interface. My problem with these very long interfaces is that they are extremely complex to document and use + there are many corner cases one could get into. Generally I think these should be really simple and try and offload the hard work to simpler methods.

> * Added tests for all different representations in 1D, 2D and 3D.

> Nick: You added the `skip_skimage` part at the top of the `test_grid.py` file. For me it skips tests even though `skimage` is available and `skip_skimage` is set to `pytest.mark.skip(False, reason="scikit-image (skimage) not available")`. Does it work properly for you?

I see, there was a bug `skip -> skipif`, added and merged, please rebase if needed.

**Comment by pfebrer** [comment]
> Couldn't we re-use the apply naming scheme to allow function arguments or strs to be fetched from numpy`?

I'm not sure I fully understand what you are proposing, but if I understand correctly, you want the value of this setting to be passed to `Grid.apply` ? The problem is that these are methods that reduce the dimensionality of the grid, so they can't be treated by the `apply` method unless we improve it. Also, to be able to generalize, we should impose the constraint that the reducing function accepts the `axis` argument, otherwise we will not be able to support everything that can be done.

I think it could be a great idea if we find a nice way to generalize dimension reduction in the `Grid.apply` method.

Did you mean something different?

**Comment by zerothi** [comment]
Ok, I am a bit lost now, we moved some of the discussion elsewhere, so where are we here? ;)

**Comment by pfebrer** [comment]
Yes my bad, I fixed some other things about the grid plot in this PR. The main thing to discuss is the order in which things are processed.

Right now it is:
- Choose representation (`real`, `imag`, ...)
- Transform the grid (i.e. apply functions such as `np.cos`).
- Cut the grid according to `x_range`, `y_range` and `z_range`.
- Interpolate values to a grid of different shape if interp is passed.
- Reduce dimensionality with `average` or `sum`.

Also, I needed you to clarify what you meant here:

> Couldn't we re-use the apply naming scheme to allow function arguments or strs to be fetched from numpy`? That would also allow more functionality without complicating the interface. My problem with these very long interfaces is that they are extremely complex to document and use + there are many corner cases one could get into. Generally I think these should be really simple and try and offload the hard work to simpler methods.

If you could give an example that would be nice :)

**Comment by zerothi** [comment]
> Right now it is:
> 
> * Choose representation (`real`, `imag`, ...)
> * Transform the grid (i.e. apply functions such as `np.cos`).
> * Cut the grid according to `x_range`, `y_range` and `z_range`.
> * Interpolate values to a grid of different shape if interp is passed.
> * Reduce dimensionality with `average` or `sum`.

I think I would just swap the 2 last ones. Interpolation and reducing dimensionality seems to be problematic. I think having more consistent numbers by averaging over *true* numbers, then interpolating would be better. Otherwise I think it looks ok.

File: sisl/viz/plotly/plots/grid.py [hunk 3]
```diff
@@ -609,6 +618,34 @@ def _transform_grid(grid, transform):
...
+    
+    @staticmethod
+    def _get_representation(grid, represent):
+        """Returns a representation of the grid
+        
+        Parameters
+        ------------
+        grid: sisl.Grid
+            the grid for which we want return
+        represent: {"real", "imag", "mod", "deg_phase", "rad_phase"}
```
Comment (zerothi) [inline sisl/viz/plotly/plots/grid.py:630]: perhaps a default `phase` which is using `radians`?

**Comment by zerothi** [comment]
> > Couldn't we re-use the apply naming scheme to allow function arguments or strs to be fetched from numpy`? That would also allow more functionality without complicating the interface. My problem with these very long interfaces is that they are extremely complex to document and use + there are many corner cases one could get into. Generally I think these should be really simple and try and offload the hard work to simpler methods.
> 
> If you could give an example that would be nice :)

My point is that your `_set_data` is extremely cumbersome to understand :) Its # of arguments is 15 which is very hard to understand and follow. While many of the flags are passed automatically, there may be better ways of doing these complicated things. I.e. not necessarily all in 1 go.

So it was more the _idea_ rather than anything now. :)

File: sisl/viz/plotly/plots/grid.py (hunk 3 above)
Comment (pfebrer) [reply sisl/viz/plotly/plots/grid.py:630]: Actually, currently anything that ends with `"phase"` and doesn't begin with `"deg"` is interpreted as `"rad_phase"`. I should probably be more strict and accept only "phase", "rad_phase" and "deg_phase", right? :sweat_smile:

Comment (zerothi) [reply sisl/viz/plotly/plots/grid.py:630]: Yeah ;)

**Comment by pfebrer** [comment]
> My point is that your _set_data is extremely cumbersome to understand :)

Totally, that's why I'm trying to put each part into a method, to improve readability. The arguments are there because these are the settings that `_set_data` uses. It's the way of telling the `Plot` class that `_set_data` should be ran if these settings are updated. Also it allows other classes to pass some settings:

https://github.com/pfebrer/sisl/blob/2cd243353f09b3611ae49d9fb99598b4ff2c8a92/sisl/viz/plotly/plots/grid.py#L1500

> I think I would just swap the 2 last ones.

Good!

**Comment by pfebrer** [comment]
Everything is ready now, all tests pass, ready to merge :+1:

**Review by zerothi — APPROVED** [review]
